src/refiner/refiner.fun

```diff
@@ -281,6 +282,61 @@ struct
         ] BY mkEvidence EQ_MEMBER_EQ
       end
 
+    fun SubtypeEq (H >> P) =
+      let
+        val #[M, N, univ] = P ^! EQ
+        val (UNIV k, #[]) = asApp univ
+        val #[L, R] = M ^! SUBTYPE
+        val #[L', R'] = M ^! SUBTYPE
+      in
+        [ H >> C.`> EQ $$ #[L, L', univ]
+        , H >> C.`> EQ $$ #[R, R', univ]
+        ] BY mkEvidence SUBTYPE_EQ
+      end
+
+     fun SubtypeMemEq (H >> P) =
+      let
+        val #[M, N, E] = P ^! EQ
+        val #[] = M ^! AX
+        val #[] = N ^! AX
+        val #[_, _] = E ^! SUBTYPE
+      in
+        [ H >> E
+        ] BY mkEvidence SUBTYPE_MEMBER_EQ
+      end
+
+    fun SubtypeIntro oname (H >> P) =
+      let
+        val #[M, N] = P ^! SUBTYPE
+        val name =
+          case oname of
+              SOME name => name
+            | NONE => (Context.fresh (H, Variable.named "x"))
+        val H' = H @@ (name, M)
```

Comment:
Hmm, I'm not sure. that's what we need to express though now that I think about it..